server/src/main/java/org/opensearch/rest/action/cat/RestIndicesAction.java

```diff
@@ -205,6 +218,91 @@ public void onFailure(final Exception e) {
         };
     }
 
+    public RestChannelConsumer doPaginatedCatRequest(
+        final RestRequest request,
+        final NodeClient client,
+        final TimeValue clusterManagerNodeTimeout,
+        final String[] indices
+    ) {
+        final boolean local = request.paramAsBoolean("local", false);
+        final boolean includeUnloadedSegments = request.paramAsBoolean("include_unloaded_segments", false);
+        final String requestedToken = request.param("next_token");
+        IndexBasedPaginationStrategy.validateRequestedRequest(requestedToken);
+        final int pageSize = Integer.parseInt(request.param("max_page_size", DEFAULT_CAT_INDICES_PAGE_SIZE_STRING));
+        final boolean latestIndicesFirst = request.paramAsBoolean("latest_indices_first", false);
+
+        return channel -> {
+            final ActionListener<Table> listener = ActionListener.notifyOnce(new RestResponseListener<Table>(channel) {
+                @Override
+                public RestResponse buildResponse(final Table table) throws Exception {
+                    return RestTable.buildResponse(table, channel);
+                }
+            });
+
+            // Fetch all the indices from clusterStateRequest for a paginated query.
+            sendClusterStateRequest(
+                indices,
+                IndicesOptions.lenientExpandHidden(),
+                local,
+                clusterManagerNodeTimeout,
+                client,
+                new ActionListener<ClusterStateResponse>() {
+                    @Override
+                    public void onResponse(final ClusterStateResponse clusterStateResponse) {
+                        IndexBasedPaginationStrategy paginationStrategy = new IndexBasedPaginationStrategy(
+                            requestedToken,
+                            pageSize,
+                            latestIndicesFirst,
+                            clusterStateResponse.getState()
+                        );
+
+                        final GroupedActionListener<ActionResponse> groupedListener = createGroupedListener(
+                            request,
+                            4,
+                            listener,
+                            new Table.PaginationMetadata(
```

Comment:
Is this required to be passed even before the table is populated ? Can it be set just before we return the response ?

Reply:
Since these actions make use of GroupedListeners, wherein we need to pre-define the logic to build table, it's kind of required to set it before the next set of Transport calls are made.